core/src/databases/database.rs

```diff
@@ -83,20 +84,210 @@ impl Database {
+                            .enumerate()
+                            .map(|(i, column_name)| {
+                                Ok((
+                                    column_name.clone(),
+                                    match row.get(i) {
+                                        Err(e) => {
+                                            return Err(anyhow!(
+                                                "Failed to retrieve value for column {}: {}",
+                                                column_name,
+                                                e
+                                            ))
+                                        }
+                                        Ok(v) => match v {
+                                            rusqlite::types::Value::Integer(i) => {
+                                                Ok(serde_json::Value::Number(i.into()))
+                                            }
+                                            rusqlite::types::Value::Real(f) => {
+                                                match serde_json::Number::from_f64(f) {
+                                                    Some(n) => Ok(serde_json::Value::Number(n)),
+                                                    None => Err(anyhow!(
+                                                        "Invalid float value for column {}",
+                                                        column_name
+                                                    )),
+                                                }
+                                            }
+                                            rusqlite::types::Value::Text(t) => {
+                                                Ok(serde_json::Value::String(t.clone()))
+                                            }
+                                            rusqlite::types::Value::Blob(b) => {
+                                                match String::from_utf8(b.clone()) {
+                                                    Err(_) => Err(anyhow!(
+                                                        "Invalid UTF-8 sequence for column {}",
+                                                        column_name
+                                                    )),
+                                                    Ok(s) => Ok(serde_json::Value::String(s)),
+                                                }
+                                            }
+                                            rusqlite::types::Value::Null => {
+                                                Ok(serde_json::Value::Null)
+                                            }
+                                        },
+                                    }?,
+                                ))
+                            })
+                            .collect::<Result<serde_json::Value>>()
+                    })?
+                    .collect::<Result<Vec<_>>>()?
```

Comment:
@spolu I don't know how to remove this `collect` -- the result of `query_and_then` doesn't have anything that allows me to "invert result" it so it looks like I have to collect as vec before I can map over it. Maybe I am missing smthg ofc

Reply:
Hard to say would have to play with it happy to make a pass at this once you're finalized